modules/ip/control/nexthop.c

```diff
@@ -32,6 +33,80 @@ struct nexthop *ip4_nexthop_lookup(uint16_t vrf_id, ip4_addr_t ip) {
 	return nexthop_lookup(nh_pool, vrf_id, &ip);
 }
 
+static control_input_t ip_output_node;
+
+void ip4_nexthop_unreachable_cb(struct rte_mbuf *m) {
+	struct rte_ipv4_hdr *ip = rte_pktmbuf_mtod(m, struct rte_ipv4_hdr *);
+	ip4_addr_t dst = ip->dst_addr;
+	struct nexthop *nh;
+
+	nh = ip4_route_lookup(control_output_mbuf_data(m)->iface->vrf_id, dst);
+	if (nh == NULL)
+		goto free; // route to dst has disappeared
+
+	if (nh->flags & GR_NH_F_LINK && dst != nh->ipv4) {
+		// The resolved nexthop is associated with a "connected" route.
+		// We currently do not have an explicit route entry for this
+		// destination IP.
+		struct nexthop *remote = ip4_nexthop_lookup(nh->vrf_id, dst);
+
+		if (remote == NULL) {
+			// No existing nexthop for this IP, create one.
+			remote = ip4_nexthop_new(nh->vrf_id, nh->iface_id, dst);
+		} else if (remote->flags & GR_NH_F_GATEWAY && remote->iface_id == 0) {
+			// Gateway route with uninitialized destination.
+			// Now, we can at least know what is the output interface.
+			remote->iface_id = nh->iface_id;
+		}
+
+		if (remote == NULL) {
+			LOG(ERR, "cannot allocate nexthop: %s", strerror(errno));
+			goto free;
+		}
+		if (remote->iface_id != nh->iface_id)
+			ABORT(IP4_F " nexthop lookup gives wrong interface", &ip);
+
+		// Create an associated /32 route so that next packets take it
+		// in priority with a single route lookup.
+		if (ip4_route_insert(nh->vrf_id, dst, 32, remote) < 0) {
+			LOG(ERR, "failed to insert route: %s", strerror(errno));
+			goto free;
+		}
+		nh = remote;
+	}
+
+	if (nh->flags & GR_NH_F_REACHABLE) {
+		// The nexthop may have become reachable while the packet was
+		// passed from the datapath to here. Re-send it to datapath.
+		struct ip_output_mbuf_data *d = ip_output_mbuf_data(m);
+		d->nh = nh;
+		if (post_to_stack(ip_output_node, m) < 0) {
+			LOG(ERR, "post_to_stack: %s", strerror(errno));
+			goto free;
+		}
+		return;
+	}
+
+	if (nh->held_pkts_num < NH_MAX_HELD_PKTS) {
```

Comment:
Just a note that holding on this packet in control path means that the data path packet pool may end up in a starvation situation. Imagine many packets being queued, with unresolved nh, and no packet left in the packet pool to receive a arp reply.
Grout would probably break from this situation later (as packets are held with a timeout), but still it would be more robust to hold on a packet copy and free the original mbuf.